### PR TITLE
Handle session timeout on client and server

### PR DESCRIPTION
**Pull request by hendrikebbers**
SessionTimeout can be configured in dolphin.properties on the server
ClientCOntext provides error handler for the remoting layer. Based on the exception type a client can check if the error is based on session timeout, bad response, etc.

---
This change is [Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/179)

### CODE REVIEW

**Comment by hendrikebbers** [comment]
Before this can be merged we need to discuss with @netopyr if we can reproduce it in JS

File: platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/impl/HttpErrorHandler.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package com.canoo.dolphin.client.impl;
+
+import com.canoo.dolphin.util.DolphinRemotingException;
+
+/**
+ * Created by hendrikebbers on 20.05.16.
+ */
+public class HttpErrorHandler {
```
Comment (hendrikebbers) [inline platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/impl/HttpErrorHandler.java]: This class can be removed

**Comment by netopyr** [comment]
Review status: 0 of 16 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/DolphinConnectionException.java, line 6 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/179#-KISpmAM4Yh_bhqaL8Wi:-KISpmAM4Yh_bhqaL8Wj:-340773127) ([raw file](https://github.com/canoo/dolphin-platform/blob/6b5423c84aa4576fc93ba863d0bc3179cf372032/platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/DolphinConnectionException.java#L6)):*
> ```Java
>  * Created by hendrikebbers on 20.05.16.
>  */
> public class DolphinConnectionException extends Exception {
> ```

Is this Exception still in use?

---

*[platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/SessionLostException.java, line 6 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/179#-KISpVEMuV9JbhUS7GpK:-KISpVEMuV9JbhUS7GpL:-1360652332) ([raw file](https://github.com/canoo/dolphin-platform/blob/6b5423c84aa4576fc93ba863d0bc3179cf372032/platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/SessionLostException.java#L6)):*
> ```Java
>  * Created by hendrikebbers on 20.05.16.
>  */
> public class SessionLostException extends Exception {
> ```

Is this Exception still being used somewhere?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/canoo/dolphin-platform/179)*

**Comment by hendrikebbers** [comment]
Review status: 0 of 16 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/DolphinConnectionException.java, line 6 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/179#-KISpmAM4Yh_bhqaL8Wi:-KISrjmIHcVcyduOVb6y:5174662) ([raw file](https://github.com/canoo/dolphin-platform/blob/6b5423c84aa4576fc93ba863d0bc3179cf372032/platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/DolphinConnectionException.java#L6)):*
<details><summary><i>Previously, netopyr (Michael Heinrichs) wrote…</i></summary>
> Is this Exception still in use?
</details>
no

---

*[platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/SessionLostException.java, line 6 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/179#-KISpVEMuV9JbhUS7GpK:-KISrj-vjjsRNAYMm83a:5174662) ([raw file](https://github.com/canoo/dolphin-platform/blob/6b5423c84aa4576fc93ba863d0bc3179cf372032/platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/SessionLostException.java#L6)):*
<details><summary><i>Previously, netopyr (Michael Heinrichs) wrote…</i></summary>
> Is this Exception still being used somewhere?
</details>
no

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/canoo/dolphin-platform/179)*

**Comment by netopyr** [comment]
I went through the changes and think we can implement them in JS as well.

**Comment by hendrikebbers** [comment]
merged master in this branch

File: platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/ClientContext.java [hunk 2]
```diff
@@ -51,4 +56,7 @@
      */
     CompletableFuture<Void> disconnect();
 
+    Subscription onRemotingError(Callback<DolphinRemotingException> callback);
```
Comment (AndreasEhret) [inline platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/ClientContext.java]: add javadoc

File: platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/DolphinConnectionException.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package com.canoo.dolphin.client;
+
+/**
+ * Created by hendrikebbers on 20.05.16.
```
Comment (AndreasEhret) [inline platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/DolphinConnectionException.java]: remove comment

File: platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/DolphinConnectionException.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ */
+public class DolphinConnectionException extends Exception {
+
```
Comment (AndreasEhret) [inline platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/DolphinConnectionException.java]: add serializable id

File: platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/DolphinSessionException.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package com.canoo.dolphin.client;
+
+/**
+ * Created by hendrikebbers on 20.05.16.
+ */
+public class DolphinSessionException extends RuntimeException {
+
```
Comment (AndreasEhret) [inline platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/DolphinSessionException.java]: add serializable id

File: platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/SessionLostException.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package com.canoo.dolphin.client;
+
+/**
+ * Created by hendrikebbers on 20.05.16.
+ */
+public class SessionLostException extends Exception {
+
```
Comment (AndreasEhret) [inline platform/dolphin-platform-client/src/main/java/com/canoo/dolphin/client/SessionLostException.java]: add serializable id

**Pull request merged**